test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/ReflectionTest.scala

```diff
@@ -131,6 +133,103 @@ class ReflectionTest {
     if (hiddenJSObj.getClass() != null)
       fail("optimizer thought that hiddenJSObj.getClass() was non-null")
   }
+
+  @Test def jsTypesKeptOnlyForTheirData_Issue4850(): Unit = {
+    import JSTypesKeptOnlyForTheirData._
+
+    @noinline
+    def nameOf(cls: Class[_]): String = cls.getName()
+
+    @inline
+    def testName(expectedShortName: String, cls: Class[_]): Unit = {
+      val prefix = "org.scalajs.testsuite.compiler.ReflectionTest$JSTypesKeptOnlyForTheirData$"
+      val expectedName = prefix + expectedShortName
+
+      assertEquals(expectedName, cls.getName()) // constant-folded
+      assertEquals(expectedName, nameOf(cls)) // evaluated at run-time
+    }
+
+    testName("NativeClass", classOf[NativeClass])
+    testName("NativeObject$", classOf[Array[NativeObject.type]].getComponentType())
```

Comment:
Is there a reason the objects/modules are always in arrays or is this just by happenstance?